rust-runtime/aws-smithy-runtime/tests/retries.rs

```diff
@@ -0,0 +1,204 @@
+}
+
+#[derive(Debug, Default)]
+struct Inner {
+    attempts: usize,
+    retry_partition: Option<String>,
+}
+
+impl OperationState {
+    fn new() -> Self {
+        OperationState {
+            inner: Arc::new(Mutex::new(Inner::default())),
+        }
+    }
+    fn attempts(&self) -> usize {
+        self.inner.lock().unwrap().attempts
+    }
+
+    fn retry_partition(&self) -> String {
+        let inner = self.inner.lock().unwrap();
+        inner
+            .retry_partition
+            .as_ref()
+            .expect("retry partition set")
+            .clone()
+    }
+}
+
+impl Intercept for OperationState {
+    fn name(&self) -> &'static str {
+        "OperationState"
+    }
+
+    fn read_before_attempt(
+        &self,
+        _context: &BeforeTransmitInterceptorContextRef<'_>,
+        _runtime_components: &RuntimeComponents,
+        cfg: &mut ConfigBag,
+    ) -> Result<(), BoxError> {
+        let mut inner = self.inner.lock().unwrap();
+        inner.attempts += 1;
+        let retry_partition = cfg
+            .load::<RetryPartition>()
+            .expect("set by default retry plugin");
+        inner.retry_partition = Some(retry_partition.to_string());
+        Ok(())
+    }
+}
+
+fn operation(
+    service: impl Into<String>,
+    max_attempts: usize,
+    http_client: impl Into<SharedHttpClient>,
+) -> (Operation<(), String, Infallible>, OperationState) {
+    #[derive(Debug)]
+    struct Deserializer;
+    impl DeserializeResponse for Deserializer {
+        fn deserialize_nonstreaming(
+            &self,
+            resp: &HttpResponse,
+        ) -> Result<Output, OrchestratorError<Error>> {
+            if resp.status().is_success() {
+                Ok(Output::erase("output".to_owned()))
+            } else {
+                Err(OrchestratorError::connector(ConnectorError::io(
+                    "mock connector error".into(),
+                )))
+            }
+        }
+    }
+
+    let attempts = OperationState::new();
+
+    let op = Operation::builder()
+        .service_name(service.into())
+        .operation_name("test")
+        .http_client(http_client.into())
+        .endpoint_url("http://localhost:1234/doesntmatter")
+        .no_auth()
+        .retry_classifier(HttpStatusCodeClassifier::default())
+        .standard_retry(
+            &RetryConfig::standard()
+                .with_max_attempts(max_attempts as u32)
+                .with_max_backoff(Duration::from_millis(1)),
+        )
+        .timeout_config(TimeoutConfig::disabled())
+        .serializer(|_body: ()| Ok(HttpRequest::new(SdkBody::empty())))
+        .deserializer_impl(Deserializer)
+        .interceptor(attempts.clone())
+        .build();
+
+    (op, attempts)
+}
+
+/// Test we exhaust the token bucket long before we exhaust max attempts
+///
+/// see [aws-sdk-rust#1234](https://github.com/awslabs/aws-sdk-rust/issues/1234)
+#[tokio::test]
+async fn token_bucket_exhausted_before_max_attempts() {
```

Comment:
I stil wish the SEP was clearer on why 50 attempts is the right number. Why 50? Why not 42 or 20 or another number?